domains/pallets/domain-registry/src/lib.rs

```diff
@@ -0,0 +1,626 @@
+
+        /// Minimal stake to be a domain operator.
+        ///
+        /// This is global, each domain can have its own minimum stake requirement
+        /// but must be no less than this value.
+        // TODO: When an executor decreases its stake in pallet-executor-registry, we should ensure
+        // the new stake amount still meets the operator stake threshold on all domains he stakes.
+        #[pallet::constant]
+        type MinDomainOperatorStake: Get<BalanceOf<Self>>;
+    }
+
+    #[pallet::pallet]
+    #[pallet::generate_store(pub(super) trait Store)]
+    #[pallet::without_storage_info]
+    pub struct Pallet<T>(_);
+
+    /// Domain configuration.
+    #[derive(Debug, Encode, Decode, TypeInfo, Clone, PartialEq, Eq)]
+    #[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
+    pub struct DomainConfig<Hash, Balance> {
+        /// Hash of the domain wasm runtime blob.
+        pub wasm_runtime_hash: Hash,
+
+        // May be supported later.
+        //pub upgrade_keys: Vec<AccountId>,
+        // TODO: elaborate this field.
+        pub bundle_frequency: u32,
+
+        /// Maximum domain bundle size in bytes.
+        pub max_bundle_size: u32,
+
+        /// Maximum domain bundle weight.
+        pub max_bundle_weight: Weight,
+
+        /// Minimum executor stake value to be an operator on this domain.
+        pub min_operator_stake: Balance,
+    }
+
+    /// Domain id for the next domain.
+    #[pallet::storage]
+    pub(super) type NextDomainId<T> = StorageValue<_, DomainId, ValueQuery>;
+
+    /// (domain_id, domain_creator, deposit)
+    #[pallet::storage]
+    pub(super) type DomainCreators<T: Config> = StorageDoubleMap<
+        _,
+        Twox64Concat,
+        DomainId,
+        Twox64Concat,
+        T::AccountId,
+        BalanceOf<T>,
+        OptionQuery,
+    >;
+
+    /// A map tracking all the non-system domains.
+    #[pallet::storage]
+    pub(super) type Domains<T: Config> =
+        StorageMap<_, Twox64Concat, DomainId, DomainConfig<T::Hash, BalanceOf<T>>, OptionQuery>;
+
+    /// (executor, domain_id, allocated_stake_proportion)
+    #[pallet::storage]
+    pub(super) type DomainOperators<T: Config> = StorageDoubleMap<
+        _,
+        Twox64Concat,
+        T::AccountId,
+        Twox64Concat,
+        DomainId,
+        Percent,
+        OptionQuery,
+    >;
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Creates a new domain with some deposit locked.
+        // TODO: proper weight
+        #[pallet::weight(10_000)]
+        pub fn create_domain(
+            origin: OriginFor<T>,
+            deposit: BalanceOf<T>,
+            domain_config: DomainConfig<T::Hash, BalanceOf<T>>,
+        ) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+
+            Self::can_create_domain(&who, deposit, &domain_config)?;
+
+            let domain_id = Self::apply_create_domain(&who, deposit, &domain_config);
```

Comment:
are you planning to handle domain_id range handling in the next pr?

These are what we discussed. Might be outdated though
0 < system_domain <=100
100 < core_domain <=1000
1000 < open_domain

Reply:
Hmm, may not be the right next one, but supposedly soon :(